### PR TITLE
Separate training and runtime attention

### PR DESCRIPTION
**Pull request by cifkao**
Should fix the issue mentioned [here](https://github.com/ufal/neuralmonkey/pull/121#issuecomment-259075579):

> The problem is that the same `Attention` object is used for constructing both the training and runtime parts of the graph, and therefore `attentions_in_time` contains tensors from both parts. Because I'm using `attentions_in_time` for visualization, I was getting images like this:
>
> tensorboard

### CODE REVIEW

**Comment by cifkao** [comment]
> The correct way will be instantiate two attention object for each way of running the decoder and collect the attentions_in_time only from one of them.

I'm not sure if I understand this correctly. Did you mean to say "one attention object for each way of running the decoder"? That will probably be a cleaner solution, but it won't make the TF computation graph any different, will it? Both `Attention` objects will be doing the same thing, except that each one will have a different `attentions_in_time`.

**Comment by cifkao** [comment]
Is this what you had in mind? It still seems to me that with regard to `CoverageAttention`, there is really no difference from my first commit.

**Comment by jlibovicky** [comment]
After the recent changes, this PR:

* introduces the notion of `Attentive` a class that has a function that is a common ancestor of all attentive encoders (note Python 3 allows multiple inheritance), creating an attention object is centralized there
* decoders don't take the pre-made `attention_object` from an encoder, rather call a `get_attention_object` method that instantiates one for them

The benefit of having the attention object like this are:
* it solves the bug from collecting attentions from both training and runtime run of a RNN decoder into single collection
* allows doing attention over the same encoder by multiple decoders.

**Comment by jlibovicky** [comment]
@cifkao, I can't add you among the reviewers, but have a look at the code and tell me whether it is what you wanted to do.

File: neuralmonkey/encoders/image_encoder.py [hunk 1]
```diff
@@ -24,7 +24,8 @@ def __init__(self, dimension, output_shape, data_id):
         self.encoded = tf.tanh(tf.matmul(self.flat, project_w) + project_b)
 
         self.attention_tensor = None
-        self.attention_object = None
+        self.attention_object_train = None
+        self.attention_object_runtime = None
```
Comment (cifkao) [inline neuralmonkey/encoders/image_encoder.py]: I guess this should inherit from `Attentive` too.

**Comment by cifkao** [comment]
I think you forgot about `image_encoder` (or we should revert my changes to it). Otherwise, LGTM.

File: neuralmonkey/encoders/attentive.py [hunk 2]
```diff
@@ -0,0 +1,25 @@
+# pylint: disable=too-few-public-methods
+class Attentive(object):
+    def __init__(self, attention_type, **kwargs):
+        self._attention_type = attention_type
+        self._attention_kwargs = kwargs
+
+        assert hasattr(self, "name")
+        assert hasattr(self, "_padding")
+        assert hasattr(self, "_attention_tensor")
```
Comment (jindrahelcl) [inline neuralmonkey/encoders/attentive.py]: Ač je to asi OK, nepřijde mi jako nejlepší věc diktovat z předka místo v `__init__` metodě potomků, ze kterýho se bude volat `super`.. Navíc mi přijde, že by to asi správně mělo bejt opačně, že by se `super` měl volat jako první command (á la Java). Nejde to tomu vrazit jako argument konstruktoru, když to potřebuje?

File: neuralmonkey/encoders/attentive.py [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+    def get_attention_object(self, runtime: bool=False):
+        # pylint: disable=no-member
+        if self._attention_type and self._attention_tensor is None:
+            raise Exception("Can't get attention: missing attention tensor.")
```
Comment (jindrahelcl) [inline neuralmonkey/encoders/attentive.py]: Dal bych `ValueError` páč to má blbou hodnotu. Nebo vytvořit v tomhle modulu nějakej specifičtější error.

File: neuralmonkey/encoders/cnn_encoder.py [hunk 4]
```diff
@@ -207,14 +215,10 @@ def rnn_encoder(inputs, last_layer_size, scope):
...
-            att_in_weights = tf.squeeze(
+            self._padding = tf.squeeze(
                 tf.reduce_prod(last_padding_masks, [1]), [2])
 
-            self.attention_object = Attention(self.attention_tensor,
-                                              scope="attention_{}".format(
-                                                  name),
-                                              dropout_placeholder=self.dropout_placeholder,
-                                              input_weights=att_in_weights)
+            super(CNNEncoder, self).__init__(attention_type)
```
Comment (jindrahelcl) [inline neuralmonkey/encoders/cnn_encoder.py]: tohle mě tu dráždí.. Zaprvý z důvodů, který jsem popsal nahoře, zadruhý, protože stačí `super().__init__(attention_type)`

File: neuralmonkey/encoders/factored_encoder.py [hunk 5]
```diff
@@ -65,21 +67,16 @@ def __init__(self, max_input_len, vocabularies, data_ids, embedding_sizes,
...
                 weight_tensor = tf.concat(
                     1, [tf.expand_dims(w, 1) for w in self.padding_weights])
 
-                self.attention_object = attention_type(
-                    self.attention_tensor,
-                    scope="attention_{}".format(self.name),
-                    dropout_placeholder=self.dropout_placeholder,
-                    input_weights=weight_tensor,
-                    max_fertility=attention_fertility)
+                super(FactoredEncoder, self).__init__(
```
Comment (jindrahelcl) [inline neuralmonkey/encoders/factored_encoder.py]: ditto

File: neuralmonkey/encoders/factored_encoder.py [hunk 6]
```diff
@@ -176,12 +171,12 @@ def _create_encoder_graph(self):
         self.attention_tensor = tf.nn.dropout(self.attention_tensor,
                                               self.dropout_placeholder)
 
-
     def feed_dict(self, dataset, train=False):
-        factors = {data_id: dataset.get_series(data_id) for data_id in self.data_ids}
+        factors = {data_id: dataset.get_series(
+            data_id) for data_id in self.data_ids}
```
Comment (jindrahelcl) [inline neuralmonkey/encoders/factored_encoder.py]: já to radši škubu až před `for`, který se pak zarovná pod `data_id`, ale proti gustu.. :-)

File: neuralmonkey/encoders/image_encoder.py (hunk 1 above)
Comment (jindrahelcl) [inline neuralmonkey/encoders/image_encoder.py]: I don't know whether we've implemented attention in this encoder. It was not used in our submission for the multimodal task, so it might be safe to ignore this for now.

File: neuralmonkey/encoders/sentence_encoder.py [hunk 7]
```diff
@@ -86,15 +87,13 @@ def __init__(self,
...
+            self._attention_tensor = self._dropout(self._attention_tensor)
 
             self.encoded = tf.concat(1, encoded_tup)
 
-            self.attention_object = attention_type(
-                self.attention_tensor, scope="attention_{}".format(name),
-                input_weights=self.padding,
-                max_fertility=attention_fertility) if attention_type else None
+        super(SentenceEncoder, self).__init__(
+            attention_type, attention_fertility=attention_fertility)
```
Comment (jindrahelcl) [inline neuralmonkey/encoders/sentence_encoder.py]: použít `super().__init__`, promyslet jak by to šlo líp bez toho

File: neuralmonkey/encoders/factored_encoder.py (hunk 6 above)
Comment (jlibovicky) [inline neuralmonkey/encoders/factored_encoder.py]: Já taky, za tohole může autotpep8, to pouštim na soubory, kde pekelně moc dlouhých řádek.

File: neuralmonkey/encoders/attentive.py (hunk 2 above)
Comment (jlibovicky) [inline neuralmonkey/encoders/attentive.py]: Javový řešení by bylo udělat tady abstraktní proměnný, co by se musely oddědit, pak by super mohlo být první volání v tom konstruktoru. Ale přišlo mi, že by se tím jenom zbytečně prodloužil už tak dlouhej kód těch enkodérů.

Comment (jlibovicky) [inline neuralmonkey/encoders/attentive.py]: Jestli myslíš, že je to tak správnější, tak já to tak s radostí přepíšu, ale sám nevím.

**Comment by jindrahelcl** [comment]
nechal bych to bejt..

2016-12-09 9:19 GMT+00:00 Jindřich Libovický <notifications@github.com>:

File: neuralmonkey/decoding_function.py [hunk 8]
```diff
@@ -24,15 +25,17 @@ def __init__(self, attention_states, scope, input_weights=None,
...
+            runtime_mode: (Optional) Indicates whether the object will be used
+                          for runtime decoding.
         """
         self.scope = scope
         self.attentions_in_time = []
         self.attention_states = attention_states
         self.input_weights = input_weights
 
-        with tf.variable_scope(scope):
+        with tf.variable_scope(scope, reuse=runtime_mode):
```
Comment (jlibovicky) [inline neuralmonkey/decoding_function.py]: A je tam vůbec nutný tohelto reuse? Když se řekne reuse o úroveň výš, tak se to přenáší i na ty vnitřní scopy, ne?
Jinak mi přijde, že by to nikdy nemělo vyrobit nový proměnný, ale vždycky spadnout.
Jak by měly vypadat ty asserty, nic rozumného mě nenapadá.

**Comment by jindrahelcl** [comment]
​prislo mi uzitecny delat assert not scope.reuse těsně před tim, než se
volá reuse_variables, abys věděl, že něco děláš zbytečně. Ale může to bejt
i schválně, tak podle toho.

Nutný tam neni, jestli se dělá o úroveň výš.​

2016-12-09 9:22 GMT+00:00 Jindřich Libovický <notifications@github.com>:

File: neuralmonkey/encoders/attentive.py (hunk 2 above)
Comment (cifkao) [inline neuralmonkey/encoders/attentive.py]: Ne ze bych moc rozumel vicenasobne dedicnosti v Pythonu, ale tohle by mohl byt problem, kdybychom chteli mit vic takovych trid, ktere by do enkoderu pridavaly nejakou funkcnost dedenim.

Nebylo by lepsi, aby `Attentive` nedelal v konstruktoru nic a inicializoval se az zavolanim nejake metody?

Jeste je moznost neresit to vubec dedicnosti.

Comment (jindrahelcl) [inline neuralmonkey/encoders/attentive.py]: Mělo by to být v konstruktoru kvůli tomu, jak se tvoří tensorflow graf z konfiguráku - neni tam moc prostoru pro volání dalších metod..
Ale mohl by to volat dekodér, kterej už dostává hotový enkodéry, to je fakt. To by pak i znamenalo, že pokud attention nepoužijeme, tak by se tahle část grafu v enkodéru ani zbytečně nevytvářela.

Comment (jindrahelcl) [inline neuralmonkey/encoders/attentive.py]: Jako že `Decoder._collect_attention_objects` by ještě na těch enkodérech volalo něco jako `if encoder.supports_attention:  encoder.create_attention_graph()`, který by vracelo `Attentive` nebo `Attention` objekt.

@jlibovicky šlo by to takhle? Změnila by se tam ta hierarchie, že místo encoder is_a attentive by to bylo encoder has_a attention

Taky to můžeme nechat na pak, aby se na tenhle pull request zbytečně nenabaloval další a další kód.

Comment (cifkao) [inline neuralmonkey/encoders/attentive.py]: Myslel jsem to tak, ze enkoder by zavolal `super`, ale to by nic neudelalo, a teprv dal v konstruktoru, az by to bylo pripravene, by zavolal `self._init_attention()`.

Attention objekt se uz tak vytvari az volanim z dekoderu, jestli se nepletu.

**Comment by jindrahelcl** [comment]
Jo, to je fakt, to by asi šlo a asi by to bylo i hezčí..

2016-12-09 11:10 GMT+00:00 Ondřej Cífka <notifications@github.com>:

File: neuralmonkey/encoders/attentive.py (hunk 2 above)
Comment (jlibovicky) [inline neuralmonkey/encoders/attentive.py]: A co by se dělalo v tom `self._init_attention()`? Teď se v tom konstruktoru jenom kontrolují ty `hasattr`. Ty když se vyhodí pryč, tak se `super()` může zavolat úplně na začátku konstruiktoru enkodéru. Důvod proč jsem to udělal takhle je, aby to spadlo už při inicializaci enkodéru, pokud by tam nevznikly správné tenzory a ne až při volání té metody `get_attention_object`, protože to se děje až v dekodéru a to by mohlo být matoucí potom při ladění konfiguráků.

**Comment by jindrahelcl** [comment]
A nemohla by ta metoda `get_attention_object` mít ty tři věci jako argument?

2016-12-09 14:59 GMT+00:00 Jindřich Libovický <notifications@github.com>:

**Comment by jlibovicky** [comment]
To by mohla, bylo by to asi 1. nepřehlednější, 2. neodstraňuje se tím to, že by to padalo při konstrukci dekodéru.  Zatím se mi asi líbí Javovský nápad: attentive bude mít astraktní properties a enkodéry je budou přetěžovat.

**Comment by jlibovicky** [comment]
Teď je tu varianta `Attentive` s abstraktními properties. Komentáře, připomínky?

File: neuralmonkey/encoders/attentive.py [hunk 9]
```diff
@@ -1,25 +1,31 @@
...
-            input_weights=self._padding,
+            scope="attention_{}".format(name),
+            input_weights=self._attention_mask,
             runtime_mode=runtime,
             **self._attention_kwargs) if self._attention_type else None
+
+    @property
+    def _attention_tensor(self):
+        """Tensor over which the attention is done."""
+        raise NotImplementedError(
```
Comment (cifkao) [inline neuralmonkey/encoders/attentive.py]: Proč nepoužít [`@abstractmethod`](https://docs.python.org/3/library/abc.html#abc.abstractmethod)?

File: neuralmonkey/encoders/attentive.py [hunk 10]
```diff
@@ -0,0 +1,31 @@
+import tensorflow as tf
+
+# pylint: disable=too-few-public-methods
+class Attentive(object):
+    def __init__(self, attention_type, **kwargs):
```
Comment (jindrahelcl) [inline neuralmonkey/encoders/attentive.py]: Chtělo by to ještě doplnit docstringy, ale už nechci zdržovat merge.

File: neuralmonkey/encoders/cnn_encoder.py [hunk 11]
```diff
@@ -202,28 +204,30 @@ def rnn_encoder(inputs, last_layer_size, scope):
...
-                    "encoder-backward")
-                # pylint: disable=redefined-variable-type
+                    last_layer_size, "encoder-backward")
                 encoder_state = tf.concat(
                     1, [encoder_state, backward_encoder_state])
 
             self.encoded = encoder_state
 
-            self.attention_tensor = \
+            self.__attention_tensor = \
```
Comment (jindrahelcl) [inline neuralmonkey/encoders/cnn_encoder.py]: dvě podtržítka? Neni to moc?
plus, backslash jde vyrefakorovat do konce řádky za `reshape(`

File: neuralmonkey/encoders/cnn_encoder.py [hunk 12]
```diff
@@ -202,28 +204,30 @@ def rnn_encoder(inputs, last_layer_size, scope):
...
-                                              scope="attention_{}".format(
-                                                  name),
-                                              input_weights=att_in_weights)
+    @property
+    def _attention_tensor(self):
+        return self.__attention_tensor
+
+    @property
+    def _attention_mask(self):
+        return self.__attention_weights
```
Comment (jindrahelcl) [inline neuralmonkey/encoders/cnn_encoder.py]: mask, weights, jedno podtřžítko dvě podtržítka, to je nějakěj bordel teda

File: neuralmonkey/encoders/sentence_encoder.py [hunk 13]
```diff
@@ -86,18 +90,21 @@ def __init__(self,
                 fw_cell, bw_cell, embedded_inputs, self.sentence_lengths,
                 dtype=tf.float32)
 
-            self.attention_tensor = tf.concat(2, outputs_bidi_tup)
-            self.attention_tensor = self._dropout(self.attention_tensor)
+            self.__attention_tensor = tf.concat(2, outputs_bidi_tup)
+            self.__attention_tensor = self._dropout(self._attention_tensor)
```
Comment (jindrahelcl) [inline neuralmonkey/encoders/sentence_encoder.py]: ditto

File: neuralmonkey/encoders/sentence_encoder.py [hunk 14]
```diff
@@ -112,16 +119,16 @@ def _create_input_placeholders(self):
         self.inputs = tf.placeholder(tf.int32, shape=[None, self.max_input_len],
                                      name="encoder_input")
 
-        self.padding = tf.placeholder(
+        self.__input_weights = tf.placeholder(
```
Comment (jindrahelcl) [inline neuralmonkey/encoders/sentence_encoder.py]: ditto

File: neuralmonkey/encoders/cnn_encoder.py (hunk 11 above)
Comment (jlibovicky) [inline neuralmonkey/encoders/cnn_encoder.py]: Dvě podrtžítka, protože private a ne protected.

File: neuralmonkey/encoders/cnn_encoder.py (hunk 12 above)
Comment (jlibovicky) [inline neuralmonkey/encoders/cnn_encoder.py]: Z dvěma podržítkama je to private field, s jedním je to property odděděná od attentive.

File: neuralmonkey/encoders/sentence_encoder.py (hunk 14 above)
Comment (jlibovicky) [inline neuralmonkey/encoders/sentence_encoder.py]: Weights přejmenuju na mask.

File: neuralmonkey/encoders/cnn_encoder.py (hunk 11 above)
Comment (jindrahelcl) [inline neuralmonkey/encoders/cnn_encoder.py]: ok

File: neuralmonkey/encoders/cnn_encoder.py (hunk 12 above)
Comment (jindrahelcl) [inline neuralmonkey/encoders/cnn_encoder.py]: ok

Comment (jindrahelcl) [inline neuralmonkey/encoders/cnn_encoder.py]: Já jsem asi pro.

File: neuralmonkey/encoders/sentence_encoder.py [hunk 15]
```diff
@@ -91,7 +91,7 @@ def __init__(self,
                 dtype=tf.float32)
 
             self.__attention_tensor = tf.concat(2, outputs_bidi_tup)
-            self.__attention_tensor = self._dropout(self._attention_tensor)
+            self.__attention_tensor = self._dropout(self.__attention_tensor)
```
Comment (jindrahelcl) [inline neuralmonkey/encoders/sentence_encoder.py]: a to před tim prošly testy?

Comment (jlibovicky) [inline neuralmonkey/encoders/sentence_encoder.py]: Co by neprošlo - akorát to šlo oklikou přes tu property místo toho, aby se volal rovnou ten atribut (měl jsem kliku, že to dělalo to samý, ale bylo to sémanticky blbě).

Comment (jindrahelcl) [inline neuralmonkey/encoders/sentence_encoder.py]: jo no jo.. hm..
